File: api/youtube.py
from apiclient.discovery import build 
api_key = "AIzaSyCWqYvv4vCZ2UiHnFuCWBgHE7lUTmsCLGo"

youtube = build("youtube", "v3", developerKey = api_key)

def get_channel_videos(channel_id):
    
    # get Uploads playlist id
    res = youtube.channels().list(id=channel_id, 
                                  part='contentDetails').execute()
    playlist_id = res['items'][0]['contentDetails']['relatedPlaylists']['uploads']
    
    videos = []
    next_page_token = None
    
    while 1:
        res = youtube.playlistItems().list(playlistId=playlist_id, 
                                           part='snippet', 
                                           maxResults=50,
                                           pageToken=next_page_token).execute()
        videos += res['items']
        next_page_token = res.get('nextPageToken')
        print(res);
        if next_page_token is None:
            break
    
    return videos

videos = get_channel_videos('UCqwGaUvq_l0RKszeHhZ5leA')

for video in videos:
    print(video['snippet']['title'])

File: src/lib/youtube.ts
const api_key = import.meta.env.VITE_API_KEY;
const channel_id = "UC3wla9xMoxDu7MIZImad1kQ";


async function initGapiClient() {
    await gapi.client
        .init({
            apiKey: api_key
        })
        .then(() => {
            console.log('Client initialized');
        });
}


export async function loadGapi(){
    gapi.load('client', initGapiClient);
}

export async function getAllVidoes(){
    await gapi.client.load("youtube", "v3");

    let res = await gapi.client.youtube.channels.list({
            id: channel_id,
            part: "contentDetails"
        })
    let playlist_id = res.result.items[0].contentDetails.relatedPlaylists.uploads
    let videos:any[] = []
    let next_page_token = "";
    
    while(next_page_token != undefined){
        let res2:any = await  gapi.client.youtube.playlistItems.list({
            playlistId: playlist_id,
            part: "snippet, contentDetails",
            maxResults: 50,
            pageToken: next_page_token
        })
        next_page_token = res2.result.nextPageToken
        videos = [...videos, ...res2.result.items]
    }
   
    return videos;

    
}

export async function getThumbnail(url: string){
    let pattern = /((http(s)?:\/\/)?)(www\.)?((youtube\.com\/)|(youtu.be\/))[\S]+/;
    let url2 = url.match(pattern);
    let video_url = url2[0];
    if(video_url.includes("watch")){
        video_url = video_url.replace("https://www.youtube.com/watch?v=", "");
    }
    else if(video_url.includes("youtube")){
        video_url = video_url.replace("https://youtu.be/", "");
    }

    await gapi.client.load("youtube", "v3");
    let res = await gapi.client.youtube.videos.list({
        part: "snippet",
        id: video_url
        
    })
    if(res.result.items.length > 0){
        return res.result.items[0].snippet.thumbnails.maxres.url
    }

    return res = "https://www.pietsmiet.de/assets/pietsmiet/brand/wordmark-plain-light-detail.svg"
   
}